torch/csrc/distributed/c10d/init.cpp

```diff
@@ -1191,12 +1224,41 @@ that adds a prefix to each key inserted to the store.
               py::call_guard<py::gil_scoped_release>());
 
   intrusive_ptr_class_<::c10d::ProcessGroupNCCL::Options>(
-      processGroupNCCL, "Options")
-      .def(py::init<>())
+      processGroupNCCL,
+      "Options",
+      processGroupOptions,
+      R"(
+ProcessGroup options for the NCCL backend
+
+Arguments:
+    timeout (timedelta, optional): Timeout for operations executed against
+            the process group. Default value equals 30 minutes. This is
+            applicable only if the environment variable ``NCCL_BLOCKING_WAIT``
+            or ``NCCL_ASYNC_ERROR_HANDLING`` is set to 1. When
+            ``NCCL_BLOCKING_WAIT`` is set, this is the duration for which the
+            process will block and wait for collectives to complete before
+            throwing an exception. When ``NCCL_ASYNC_ERROR_HANDLING`` is set,
+            this is the duration after which collectives will be aborted
+            asynchronously and the process will crash. `
+    is_high_priority_stream (bool, optional): flag to enable/disable process
+            group to pick up high priority cuda streams. It lets CUDA driver
+            to prioritize NCCL kernels when there are compute kernels waiting.
+
+Example::
+    >>> import torch.distributed as dist
+    >>> from datetime import timedelta
+    >>>
+    >>> nccl_options = dist.ProcessGroupNCCL.Options(timeout=timedelta(seconds=300))
+    >>> # initialize a nccl process group with the options just created
+    >>> dist.init_process_group("nccl", pg_options=nccl_options)
+      )")
+      .def(
+          py::init<std::chrono::milliseconds, bool>(),
+          py::arg("timeout") = kProcessGroupDefaultTimeout,
```

Comment:
Why do we have a `timeout` arg in ProcessGroupOptions if both `init_process_group` and `new_group` already have a timeout?

Reply:
I think mostly from an intuition perspective, a timeout is actually an option to construct the process group, and I think it might make some sense to group all the options for ProcessGroup together in a single struct. ProcessGroupNCCL and ProcessGroupGloo both have it, so I think we might just put this into the base class and expose to the user. Do you mean that we better find a way to hide the `timeout` arg from the user and set it internally?

Reply:
I think we should have it in the options as it's an option to build the process group, but agreed that we should figure out the precedence/how to handle when both values are set.

Reply:
Sorry about the delay here, but I don't think we should have two timeouts in two different places in the API. This is confusing for users. Is the goal of this feature to retrieve the options from an existing pg and use those in `new_group` to create a new group with the same options? If so, shouldn't `pg_options` be an argument only in `new_group` and not in `init_process_group`?

Can we maybe assert that `timeout` and `pg_options` should not be specified together and the user should use only one of them?

I guess the same applies to `backend` as well? That option is now present in the API and in pg_options as well. 

cc @rohan-varma @wanchaol

Reply:
Yeah agree that we should have a unified place to put in timeout. The goal of this feature is to enable the options to be available during the creation of process group, I think both `init_process_group` and `new_group` should have this supported? since some user might only want one processgroup but configure it with the options during initialization (i.e. user only creates a nccl processgroup with high_priority_stream). Given that we don't want to confuse user for setting different timeouts, the current approach is that if we detect that user pass in timeouts and specify timeouts in ProcessGroupOptions, we error out to the user about this behavior about timeout conflicting. 

I think assert timeout and pg_options be not specified together sounds good to me, it's actually a much more clean approach. The only question is timeout is not an optional argument as of now (it uses default value), so we couldn't know if two are both specified or not (we can check if timeout uses default_pg_timeout though but that will not totally avoid it).  Let me try changing timeout to `Optional[timedelta]` in both `init_process_group` and `new_group`, I think it should be fine as we can specify default_pg_timeout inside the APIs (should not be BC breaking).